### PR TITLE
[Serve] Fix Flask Request self reference

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
It turns out there is a simpler way to solve  https://github.com/ray-project/ray/issues/12395

Benchmark didn't change:
```
Before:
Latency
50%        3.841182
99%        4.529426
Throughput
Requests/sec:   2252.57

After:
Latency
50%        3.773991
99%        4.505709
Throughput
Requests/sec:   2281.46
```

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
@edoakes in case you are interested, here's what shallow do:
https://github.com/pallets/werkzeug/blob/daeb2d0f6c911332ffbf17318553afc37b232d3a/src/werkzeug/wrappers/base_request.py#L159-L168

File: python/ray/serve/tests/test_regression.py [hunk 1]
```diff
@@ -35,6 +38,26 @@ async def __call__(self, _request):
...
+        return len(gc.get_objects())
+
+    client.create_backend("model", gc_tracked_objects)
+    client.create_endpoint("model", backend="model", route="/model")
+
+    tracked_objects: List[int] = []
+    for _ in range(10):
+        result = requests.get("http://127.0.0.1:8000/model")
+        assert result.status_code == 200
+        tracked_objects.append(result.json())
```
Comment (edoakes) [inline python/ray/serve/tests/test_regression.py:55]: please test with serve handle interface too

File: python/ray/serve/tests/test_regression.py [hunk 2]
```diff
@@ -35,6 +38,26 @@ async def __call__(self, _request):
...
+
+    # gc tracked object shouldn't grow
+    assert tracked_objects[5] == tracked_objects[-1]
```
Comment (edoakes) [inline python/ray/serve/tests/test_regression.py:58]: This looks somewhat brittle, let's just check that there aren't any unreachable references (that's what we really care about):

```
``` 
import gc
gc.set_debug(gc.DEBUG_SAVEALL)
gc.collect()
assert len(gc.garbage) == 0
```

File: python/ray/serve/http_util.py [hunk 3]
```diff
@@ -11,7 +11,10 @@ def build_flask_request(asgi_scope_dict, request_body):
     happen.
     """
     wsgi_environ = build_wsgi_environ(asgi_scope_dict, request_body)
-    return flask.Request(wsgi_environ)
+    # We set populate_request=False to prevent self reference, which leads to
```
Comment (edoakes) [inline python/ray/serve/http_util.py:14]: What does setting populate_request=False do? Any other side effects?

Comment (simon-mo) [reply python/ray/serve/http_util.py:14]: It's only used for debugging.
```
    Per default the request object will be added to the WSGI
    environment as `werkzeug.request` to support the debugging system.
```

**Review by edoakes — APPROVED** [review]

**Review by edoakes — APPROVED** [review]

**Pull request merged**